### PR TITLE
add timesync

### PR DESCRIPTION
**Pull request by fmina**
@jwidauer Here is the time sync driver. We should merge also @ds-github  rgb8-mode.

### CODE REVIEW

File: pointgrey_camera_driver/include/pointgrey_camera_driver/PointGreyCamera.h [hunk 1]
```diff
@@ -194,6 +197,12 @@ class PointGreyCamera
   /// GigE packet delay:
   unsigned int packet_delay_;
 
+  //time sync
+  FlyCapture2::TimeStamp last_timestamp_;
+  FlyCapture2::TimeStamp comulative_timestamp_;
```
Comment (rikba) [inline pointgrey_camera_driver/include/pointgrey_camera_driver/PointGreyCamera.h]: > cumulative

File: pointgrey_camera_driver/include/pointgrey_camera_driver/PointGreyCamera.h [hunk 2]
```diff
@@ -44,7 +44,10 @@ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 
 // FlyCapture SDK from Point Grey
 #include "flycapture/FlyCapture2.h"
+#include <flycapture/FlyCapture2Defs.h>
```
Comment (rikba) [inline pointgrey_camera_driver/include/pointgrey_camera_driver/PointGreyCamera.h]: Use quotation marks to meet their style?

File: pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#pragma once
```
Comment (rikba) [inline pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h:1]: What is this for?

File: pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  TimeSyncEKF();
+  ~TimeSyncEKF();
+  double getLocalTimestamp(double device_time);
+  void updateFilter(double device_time, double local_time);
+  void print();
+  void reset();
+
+ private:
+  void initialize(double device_time, double local_time);
+  bool isOutlier(double device_time, double local_time);
```
Comment (rikba) [inline pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h:28]: Is this implemented?

File: pointgrey_camera_driver/src/PointGreyCamera.cpp [hunk 5]
```diff
@@ -979,10 +979,32 @@ void PointGreyCamera::grabImage(sensor_msgs::Image &image, const std::string &fr
...
 
-    // Set header timestamp as embedded for now
-    TimeStamp embeddedTime = rawImage.GetTimeStamp();
-    image.header.stamp.sec = embeddedTime.seconds;
-    image.header.stamp.nsec = 1000 * embeddedTime.microSeconds;
+
+    TimeStamp currentTime = rawImage.GetTimeStamp();
+    int delta_cycle_seconds = currentTime.cycleSeconds - last_timestamp_.cycleSeconds;
+    if(delta_cycle_seconds < 0){
+      delta_cycle_seconds += 128;
```
Comment (rikba) [inline pointgrey_camera_driver/src/PointGreyCamera.cpp]: Can you make these numbers more verbose, e.g., `const int kCycleSecondsHalfPeriod = 128;`?

File: pointgrey_camera_driver/src/PointGreyCamera.cpp [hunk 6]
```diff
@@ -979,10 +979,32 @@ void PointGreyCamera::grabImage(sensor_msgs::Image &image, const std::string &fr
...
+    }
+
+    int delta_cycle_count = currentTime.cycleCount - last_timestamp_.cycleCount;
+    if(delta_cycle_count < 0){
+      delta_cycle_count += 8000;
```
Comment (rikba) [inline pointgrey_camera_driver/src/PointGreyCamera.cpp]: Maybe more verbosity, see above.

File: pointgrey_camera_driver/src/PointGreyCamera.cpp [hunk 7]
```diff
@@ -979,10 +979,32 @@ void PointGreyCamera::grabImage(sensor_msgs::Image &image, const std::string &fr
...
+    }
+
+    comulative_timestamp_.microSeconds += delta_cycle_count*125; //will go over 1e6
```
Comment (rikba) [inline pointgrey_camera_driver/src/PointGreyCamera.cpp]: Also maybe explain what the `125` stands for.

File: pointgrey_camera_driver/src/PointGreyCamera.cpp [hunk 8]
```diff
@@ -979,10 +979,32 @@ void PointGreyCamera::grabImage(sensor_msgs::Image &image, const std::string &fr
...
+    comulative_timestamp_.seconds += delta_cycle_seconds;
+    comulative_timestamp_.microSeconds = comulative_timestamp_.microSeconds%1000000;
```
Comment (rikba) [inline pointgrey_camera_driver/src/PointGreyCamera.cpp]: Rename `1000000` to `const int kNumMicroSecondsInSecond = 1000000;`?

**Review by rikba — APPROVED** [review]
Maybe some more verbosity in converting the device time stamp to seconds. Otherwise LGTM :)

File: pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h (hunk 4 above)
Comment (fmina) [reply pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h:28]: No

File: pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h (hunk 3 above)
Comment (fmina) [reply pointgrey_camera_driver/include/pointgrey_camera_driver/TimeSyncEKF.h:1]: to avoid multiple inclusions of the same header

**Comment by jwidauer** [comment]
@fmina Awesome! Thank you!

**Pull request merged**